src/raftstore/store/util.rs

```diff
@@ -177,22 +180,151 @@ pub fn get_region_approximate_size(db: &DB, region: &metapb::Region) -> Result<u
     Ok(size)
 }
 
+/// Lease records an expired time, for examining the current moment is in lease or not.
+/// It's dedicated to the Raft leader lease mechanism, contains either state of
+///   1. Lowerbound Timestamp
+///      An unsafe leader lease timestamp, which marks the leader may still hold or lose
+///      its lease until the clock time goes over this timestamp. It is critical that a Lease
+///      can only transit it's state to the Upperbound after the clock time goes over
+///      the Lowerbound.
+///   2. Upperbound Timestamp
+///      A safe leader lease timestamp, which marks the leader holds the lease for now.
+///      The lease is safe until the clock time goes over this timestamp.
+///
+/// ```text
+/// Time
+/// |--------------------------------------------------->
+///      ^               ^                  ^
+///     Now          Lowerbound         Upperbound
+///      |    Suspect    |      Vailed      |   Expired
+/// ```
+///
+/// Note:
+///   - Upperbound would increase when raft log entries are applied in current term.
+///   - Lowerbound would be set after the message `MsgTimeoutNow` is sent by current peer.
+///     The message `MsgTimeoutNow` starts a leader transfer procedure. During this procedure,
+///     current peer as an old leader may still hold its lease or lose it.
+///     It's possible there is a new leader elected and current peer as an old leader
+///     doesn't step down due to network partition from the new leader. In that case,
+///     current peer lose its leader lease.
+///     Within this unsafe leader lease expire time, read requests could not be performed
+///     locally.
+// TODO: add a remote Lease. A special lease that derives from Lease, it will be sent
+//       to the local read thread, so name it remote. If Lease expires, the remote must
+//       expire too.
+pub struct Lease {
+    // A lowerbound is in the Either::Left(_),
+    // an upperbound is in the Either::Right(_).
+    bound: Option<Either<Timespec, Timespec>>,
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum LeaseState {
+    Suspect,
+    Vailed,
```

Comment:
Valid?

Reply:
or can you explain these?

Reply:
Oh, It's a typo, I will add some comments for each variant.